server/src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/random/RandomSamplerAggregator.java

```diff
@@ -107,11 +108,15 @@ public void collect(int doc, long owningBucketOrd) throws IOException {
         final DocIdSetIterator docIt = scorer.iterator();
         final Bits liveDocs = ctx.reader().getLiveDocs();
         // Iterate every document provided by the scorer iterator
-        for (int docId = docIt.nextDoc(); docId != DocIdSetIterator.NO_MORE_DOCS; docId = docIt.nextDoc()) {
-            // If liveDocs is null, that means that every doc is a live doc, no need to check if it has been deleted or not
-            if (liveDocs == null || liveDocs.get(docIt.docID())) {
-                collectBucket(sub, docIt.docID(), 0);
+        try {
+            for (int docId = docIt.nextDoc(); docId != DocIdSetIterator.NO_MORE_DOCS; docId = docIt.nextDoc()) {
+                // If liveDocs is null, that means that every doc is a live doc, no need to check if it has been deleted or not
+                if (liveDocs == null || liveDocs.get(docIt.docID())) {
+                    collectBucket(sub, docIt.docID(), 0);
+                }
             }
+        } catch (CollectionTerminatedException e) {
+            return LeafBucketCollector.NO_OP_COLLECTOR;
```

Comment:
Sometimes, `collectBuckets` being called directly like this (instead of within a leaf collector called by the searcher) can throw if the underlying collector terminates (like composite on sorted indices).

Wrapping this allows sub collectors to terminate and for sampler to return early.